src/caffeine/allocation_s.f90

```diff
@@ -118,36 +126,45 @@
         call caf_deallocate(current_team%info%heap_mspace, c_loc(coarray_handles(i)%info))
     end do
     if (present(stat)) stat = 0
+    if (caf_have_child_teams()) then
+      ! reclaim any free space possible for the child teams to use
+      if (caf_this_image(current_team%info%gex_team) == 1) then
+        call caf_deallocate(current_team%info%heap_mspace, current_team%info%child_heap_info%allocated_memory)
+      end if
+      call caf_establish_child_heap
+    end if
   end procedure
 
   module procedure prif_deallocate
     call caf_deallocate(non_symmetric_heap_mspace, mem)
   end procedure
 
-  subroutine add_to_team_list(current_team, coarray_handle)
-    type(prif_team_type), intent(inout) :: current_team
+  subroutine add_to_team_list(coarray_handle)
     type(prif_coarray_handle), intent(inout) :: coarray_handle
```

Comment:
Yeah, I keep struggling with this conceptually. Sure the pointer isn't changing, but the data it points to is, so my intuition defaults to `inout`. You're right though, since the pointer isn't changing it can be `intent(in)`.